course_discovery/apps/edx_elasticsearch_dsl_extensions/distinct_counts/query.py

```diff
@@ -0,0 +1,242 @@
+            # Use the DistinctCountsElasticsearchQueryWrapper to execute the query so that distinct hit and query
+            # counts may be computed.
+
+            self._distinct_result_count = getattr(self._response, 'distinct_hits', 0)
+
+        return self._response
+
+    def validate(self):
+        """ Verify that all `FacetedSearch` options are valid and supported by this custom `FacetedSearch` class."""
+        dicted_aggs = self.aggs.to_dict().get('aggs')
+        if dicted_aggs:
+            for agg_name, options in dicted_aggs.items():
+                aggs = options.get('aggs')
+                if aggs:
+                    for field, agg_options in aggs.items():
+                        if 'date_histogram' in agg_options.keys():
+                            raise RuntimeError('DistinctCountsSearchQuerySet does not support date facets.')
+                        self._validate_field_facet_options(field, agg_options)
+
+        if self.aggregation_key is None:
+            raise RuntimeError('aggregation_key is required.')
+
+    def _validate_field_facet_options(self, field, options):
+        """ Verify that the provided field facet options are valid and can be converted to an aggregation."""
+        supported_options = DistinctCountsElasticsearchQueryWrapper.SUPPORTED_FIELD_FACET_OPTIONS
+        options_ = list(options.values())[0]
+        for option, __ in options_.items():
+            if option not in supported_options:
+                msg = (
+                    'DistinctCountsSearchQuerySet only supports a limited set of field facet options.'
+                    'Field: {field}, Supported Options: ({supported}), Provided Options: ({provided})'
+                ).format(field=field, supported=','.join(supported_options), provided=','.join(options.keys()))
+                raise RuntimeError(msg)
+
+    # pylint: disable=arguments-differ
+    @classmethod
+    def from_dict(cls, *args, **kwargs):
+        """ Raise an exception since we do not currently want/need to support raw queries."""
+        raise RuntimeError('DistinctCountsSearchQuerySet does not support raw queries.')
+
+    # pylint: disable=arguments-differ
+    def update_from_dict(self, *args, **kwargs):
+        """ Raise an exception since we do not currently want/need to support raw queries."""
+        raise RuntimeError('DistinctCountsSearchQuerySet does not support raw queries.')
+
+    def distinct_count(self):
+        """
+        Return the distinct hit count.
+        """
+
+        if self._distinct_result_count is None:
+            self.execute()
+        return self._distinct_result_count
+
+    def facet_counts(self):
+        """
+        Return the facet counts.
+        """
+        response = self.execute()
+
+        return response.facets
+
+
+class DistinctCountsElasticsearchQueryWrapper:
+    """
+    Custom search-like class that enables the computation of distinct hit and facet counts during search queries.
```

Comment:
please split the title and body of the docstrings.

Reply:
Did it.